### PR TITLE
make pre build events post build events work using targets

### PR DESCRIPTION
**Pull request by jmarolf**
**Customer scenario**

User wants to have their pre-build and post build events use MSBuild variables but they are evaluated as empty strings without this fix

**Bugs this fixes:** 

https://github.com/dotnet/project-system/issues/1569
https://github.com/dotnet/project-system/issues/1879

**Workarounds, if any**
User can manually modify their project file.

**Risk**

this only affects the project file if the user writes pre or post build events from the properties pages.  risk is low.

**Performance impact**

Code is only called when the user changes the value in the property page.  Risk is low.

**Is this a regression from a previous update?**

It is a regression from the native project system

**Root cause analysis:**

Targets are now implicitly imported to MSBuild variables do not exist.  The fix is to add a target to the project file so that when the user add pre and post build events, by the time they are evaluated the MeBuild variables exist.

**How was the bug found?**

customer reported - This is one of the top VS Feedback reports with 20+ votes.

### CODE REVIEW

**Comment by jmarolf** [comment]
@dotnet/project-system please review.  Let me know if you have a better idea for project file format.

**Comment by srivatsn** [comment]
Tagging @rainersigwald @andygerlicher @cdmihai for ideas as well. The problem here is that prebuildevent and postbuildevent used to be specificed after import of csharp.targets so that variables like $(OutDir) can be used in the events. with the SDK attribute there's no easy way to do that.

@jmarolf here is making the property pages spit out a 

```xml
<Target Name="PreBuild" BeforeTargets="PreBuildEvent">
   <Exec Command="echo $(OutDir)" />
</Target>
```

to the project file but that's not the best experience. Any better ideas?

@davkean suggested that atleast it should be 

```xml
<Target Name="PreBuild" BeforeTargets="PreBuildEvent">
      <PreBuildEvent>echo $(OutDir)</PreBuildEvent>
</Target>
```

This will require changes to msbuild though because the PreBuildEvent target is conditioned on the existence of the PreBuildEvent property.

**Comment by rainersigwald** [comment]
> but that's not the best experience

Can you elaborate, @srivatsn? I like it a lot better. As the original reasoning here reveals, defining properties that get fed into an exec task later has . . . confusing behavior.

If you really love the current behavior, you could rewrite the project to have an explicit imports instead of implicit, and add the property definition in the old place.

**Comment by jmarolf** [comment]
We can do this:

```XML
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp1.1</TargetFramework>
    <RunPostBuildEvent>Always</RunPostBuildEvent>
  </PropertyGroup>

  <Target Name="PreBuild" BeforeTargets="PreBuildEvent">
    <Exec Command="echo $(ProjectDir)" />
  </Target>

  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="echo $(ProjectDir)" />
  </Target>

</Project>
```

or this:
```XML
<Project>

  <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.props" />
  
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp1.1</TargetFramework>
    <RunPostBuildEvent>Always</RunPostBuildEvent>
  </PropertyGroup>

  <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.targets" />

  <PropertyGroup>
    <PreBuildEvent>echo $(ProjectDir)</PreBuildEvent>
    <PostBuildEvent>echo $(ProjectDir)</PostBuildEvent>
  </PropertyGroup>
  
</Project>
```
Both of which add six lines.  I chose the first because the complexities of deciding when to change implicit/explicit target and props imports seemed like a drawback.  Any other option would need to add special target or msbuild logic.  Is that what we want to do?

**Comment by jmarolf** [comment]
@srivatsn I think this PR needs to handle
1.  Converting the old property syntax to the new one
2. Persisting the explicit imports style is a user enters that into the project file manually

Anything else?

**Comment by rainersigwald** [comment]
@jmarolf Does MSBuild not do the latter already? Seems like we should.

**Comment by jmarolf** [comment]
@rainersigwald MSBuild does, I am talking about when the user navigates to the property page to set this value.  The property page should know this alternate format and not change it.

**Comment by srivatsn** [comment]
I don't like unrolling the SDK attribute for the questions mentioned above - someone adds a prebuilevent and removes it, now we have to move it back to a Project SDK attribute but only if users haven't edit the csproj and add more stuff above\below the imports. The targets solution is better than the unrolling but it's more verbose and so I was just wondering if there was something better here (assuming we can change whatever we want in the core targets)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties.InterceptedProjectProperties
+{
+    internal abstract partial class AbstractBuildEventValueProvider
+    {
+        public abstract class Helper
+        {
+            private const string _execTaskName = "Exec";
+            private const string _commandString = "Command";
+
+            protected Helper(string buildEventString,
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: `Helper` is a very generic name. This object contains state so I would suggest a better name, like `AbstractBuildEventHelper` or `AbstractBuildEventWorker`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+using Microsoft.Build.Construction;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties.InterceptedProjectProperties
+{
+    internal abstract partial class AbstractBuildEventValueProvider
+    {
+        public abstract class Helper
+        {
+            private const string _execTaskName = "Exec";
+            private const string _commandString = "Command";
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: nit: `String` in the variable name seems redundant

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+using System.Linq;
+using Microsoft.Build.Construction;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties.InterceptedProjectProperties
+{
+    internal abstract partial class AbstractBuildEventValueProvider
+    {
+        public abstract class Helper
+        {
+            private const string _execTaskName = "Exec";
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: nit: just `ExecTask` maybe?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                   Func<ProjectTargetElement, string> getTargetString,
+                   Action<ProjectTargetElement> setTargetDependencies)
+            {
+                BuildEventString = buildEventString;
+                TargetNameString = targetNameString;
+                GetTargetString = getTargetString;
+                SetTargetDependencies = setTargetDependencies;
+            }
+
+            private Func<ProjectTargetElement, string> GetTargetString { get; }
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: nit: I am not too bothered about the redundant naming here. But just wanted to bring it to your notice .

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                    SetParameter(projectXml, unevaluatedPropertyValue);
+                }
+            }
+
+            private (bool success, ProjectTaskElement execTask) FindExecTaskInTargets(ProjectRootElement projectXml)
+            {
+                var execTask = projectXml.Targets
+                                    .Where(target => GetTargetString(target) == BuildEventString)
+                                    .SelectMany(target => target.Tasks)
+                                    .Where(task => task.Name == _execTaskName)
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: Msbuild syntax itself is case-insensitive. We should account for it.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                else
+                {
+                    SetParameter(projectXml, unevaluatedPropertyValue);
+                }
+            }
+
+            private (bool success, ProjectTaskElement execTask) FindExecTaskInTargets(ProjectRootElement projectXml)
+            {
+                var execTask = projectXml.Targets
+                                    .Where(target => GetTargetString(target) == BuildEventString)
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: case-insensitive comparison required

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            }
+
+            private Func<ProjectTargetElement, string> GetTargetString { get; }
+            private Action<ProjectTargetElement> SetTargetDependencies { get; }
+            private string BuildEventString { get; }
+            private string TargetNameString { get; }
+
+            public string GetProperty(ProjectRootElement projectXml)
+            {
+                var result = FindExecTaskInTargets(projectXml);
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: why return a tuple here? Instead, cant we do a null check to determine success?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                {
+                    return null;
+                }
+
+                if (result.execTask.Parameters.TryGetValue(_commandString, out var commandText))
+                {
+                    return commandText;
+                }
+
+                return null; //unreachable
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: Better to add a Assert fail so that we get notified when this happens rather than silently passing

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+            private (bool success, ProjectTargetElement target) FindTargetToRemove(ProjectRootElement projectXml)
+            {
+                var targetArray = projectXml.Targets
+                                        .Where(target =>
+                                            GetTargetString(target) == BuildEventString &&
+                                            target.Children.Count == 1 &&
+                                            target.Tasks.Count == 1 &&
+                                            target.Tasks.First().Name == _execTaskName);
+                return (success: targetArray.Count() == 1, target: targetArray.SingleOrDefault());
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: It does not make sense to have the same target twice but nevertheless we do allow it. If there is more than one, then we will crash since we are using `Single`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+                                    .Where(task => task.Name == _execTaskName)
+                                    .FirstOrDefault();
+                return (success: execTask != null, execTask: execTask);
+            }
+
+            private (bool success, ProjectTargetElement target) FindTargetToRemove(ProjectRootElement projectXml)
+            {
+                var targetArray = projectXml.Targets
+                                        .Where(target =>
+                                            GetTargetString(target) == BuildEventString &&
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: case insensitive comparison here as well

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+                    SetExecParameter(execTask, unevaluatedPropertyValue);
+                }
+            }
+
+            private void SetExecParameter(ProjectTaskElement execTask, string unevaluatedPropertyValue)
+                => execTask.SetParameter(_commandString, unevaluatedPropertyValue);
+
+            private string GetTargetName(ProjectRootElement projectXml)
+            {
+                var targetNames = projectXml.Targets.Select(t => t.Name).ToArray();
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.Helper.cs]: We are going to allocate a string for all targets. `.Contains` on the array wont be efficient as well. Instead if we use Hashset then the search is a constant operation

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs [hunk 12]
```diff
@@ -0,0 +1,524 @@
...
+            var expected = @"<?xml version=""1.0"" encoding=""utf-16""?>
+<Project Sdk=""Microsoft.NET.Sdk"">
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp1.1</TargetFramework>
+  </PropertyGroup>
+  <Target Name=""PostBuild"">
+  </Target>
+  <Target Name=""PostBuild1"" AfterTargets=""PostBuildEvent"">
+    <Exec Command=""echo &quot;post build $(OutDir)&quot;"" />
```
Comment (rainersigwald) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: This is pretty awesome. Is there equivalent removal logic?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs [hunk 13]
```diff
@@ -0,0 +1,524 @@
...
+            var actual = systemUnderTest.GetProperty(root);
+            Assert.Equal(@"echo $(ProjectDir)", actual);
+        }
+
+        [Fact]
+        public static void SetPropertyTest_NoTargetsPresent()
+        {
+            var root = @"
+<Project Sdk=""Microsoft.NET.Sdk"">
+
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: What happens to all the line breaks? Seems like we are modifying the format of the project file. This could be an issue unrelated to your change. Are we tracking it anywhere? If not, we should

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs [hunk 14]
```diff
@@ -0,0 +1,524 @@
...
+    <TargetFramework>netcoreapp1.1</TargetFramework>
+  </PropertyGroup>
+</Project>";
+
+            var actual = stringWriter.ToString();
+            Assert.Equal(expected, actual);
+        }
+
+        [Fact]
+        public static void SetPropertyTest_RemoveTarget_NewlineCharacter()
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: We are not removing the target in this testcase. What is the expected behavior when the property is set to `\r\n`?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs [hunk 15]
```diff
@@ -0,0 +1,524 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.ProjectSystem.VS.Properties.InterceptedProjectProperties;
+using Microsoft.VisualStudio.ProjectSystem.VS.Utilities;
+using Xunit;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties
+{
+    [ProjectSystemTrait]
+    public class PostBuildEventValueProviderTests
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: We need some tests to show that we handle case insensitivity as well

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PreBuildEventValueProviderTests.cs [hunk 16]
```diff
@@ -0,0 +1,524 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.ProjectSystem.VS.Properties.InterceptedProjectProperties;
+using Microsoft.VisualStudio.ProjectSystem.VS.Utilities;
+using Xunit;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties
+{
+    [ProjectSystemTrait]
+    public class PreBuildEventValueProviderTests
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PreBuildEventValueProviderTests.cs]: Since the tests are a mirror of the PostBuildEvent, all the comments, above, apply here as well

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+                if (result.success)
+                {
+                    return result.property;
+                }
+
+                // Check if build events can be found in targets
+                return GetFromTargets(projectXml);
+            }
+
+            public void SetProperty(string unevaluatedPropertyValue, ProjectRootElement projectXml)
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs]: Get and Set Property are doing different things. Get does not care if there are more than one Task and always uses the First Task. Set, on the other hand, when trying to remove, looks for the target with just one Exec task. 

This behavior accomadates the common scenario. But we need to handle the other scenario of having more than one task because this behavior silently passes without telling the user something was wrong.

Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs]: We need not address this issue in this PR. We can open an issue to track this though.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs (hunk 15 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: Unless I am missing something `ProjectLockAwaitable` is un-mockable being a struct. Can you point me to other classes that use `ReadLockAsync()` and `GetProjectXmlAsync()`?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs (hunk 12 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: I'll add a test

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs (hunk 13 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: `root.Save(TextWriter)` and `root.Save(string path)` use completely different code paths.  I'll change the tests to write out a file so the newlines are not changed

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+            private Func<ProjectTargetElement, string> GetTarget { get; }
+            private Action<ProjectTargetElement> SetTargetDependencies { get; }
+            private string BuildEvent { get; }
+            private string TargetName { get; }
+
+            public string GetProperty(ProjectRootElement projectXml)
+            {
+                // Check if project file already has props in place for this.
+                var result = TryGetFromProperty(projectXml);
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs]: I don't like the pattern of a method returning null to indicate failure.  `null` is a valid thing to return from `GetProperty` itself.  Someone could mistakenly assume that they should return null if one of the methods they call gives them null, instead of continuing.  Roslyn uses [Optional](http://source.roslyn.io/#Microsoft.CodeAnalysis/Optional.cs,86bbd87a0ffc79bf,references) to indicate whether a returned value is valid, but this pattern was established before tuples.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs (hunk 14 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: In the old project system, newlines were persisted but spaces or tabs were not.  I am preserving this behavior

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs (hunk 17 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs]: My logic is:
- if the pre-build or post-build target has multiple exec tasks  there is no way for us to know which one to use so we select the first one.  That way we can at least make the behavior observably deterministic
- If the user deletes the text of a post or pre build event (or sets it to just spaces and tabs) we remove the target only if there is only one exec task.  This ensures that we do not remove targets that have custom logic written by the developer.

Both cases, the user has written something custom so we err on the side of not changing things.  What do we need to communicate to the user in the scenario:
> But we need to handle the other scenario of having more than one task because this behavior silently passes without telling the user something was wrong.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs (hunk 12 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: test has been added

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs (hunk 15 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: done

**Comment by jmarolf** [comment]
@balajikris can you do another review pass? thanks!

**Comment by davkean** [comment]
@jmarolf Can you summarize the new behavior?

**Comment by jmarolf** [comment]
@davkean here are the basics.  I can write a more formal doc. 

1. If the user has `PreBuildEvent` or `PostBuildEvent` properties defined already in their project we use those properties and do not attempt to convert them.  So this proj file is not modified even though the msbuild variable `ProjectDir` will not be correctly evaluated

```XML
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp1.1</TargetFramework>
    <PreBuildEvent>echo $(ProjectDir)</PreBuildEvent>
    <PostBuildEvent>echo $(ProjectDir)</PostBuildEvent>
  </PropertyGroup>
</Project>
```
image

2. If the user has no `PreBuildEvent` or `PostBuildEvent` properties defined we will add a target to their project with an exec task when they modify the build events tab.

So this:
```XML
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp1.1</TargetFramework>
  </PropertyGroup>
</Project>
``` 

Becomes this:
```XML
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp1.1</TargetFramework>
  </PropertyGroup>
  <Target Name="PreBuild" BeforeTargets="PreBuildEvent">
    <Exec Command="echo $(ProjectDir)" />
  </Target>
  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="echo $(ProjectDir)" />
  </Target>
</Project>
``` 

When the build events tab looks like this:
image

On a new project

3. In the event that the user has a target defined in their project that does not match 

This for prebuild
```XML
<Target Name="PreBuild" BeforeTargets="PreBuildEvent">
  <Exec Command="echo $(ProjectDir)" />
</Target>
```

Or this for postbuild:
```XML
<Target Name="PostBuild" AfterTargets="PostBuildEvent">
  <Exec Command="echo $(ProjectDir)" />
</Target>
```

We will add a new target with a unique name (Post/Pre Build + N) whenever they edit the build events tab

4. If the developer deletes their text from the build events tab we will remove the targets iff they match the rules for 3

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs (hunk 14 above)
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/PostBuildEventValueProviderTests.cs]: ok. Rename the test name accordingly. Test still says `RemoveTarget`

**Comment by tannergooding** [comment]
So, if you have two targets with the same 'BeforeTargets':
```xml
  <Target Name="PreBuild1" BeforeTargets="PreBuildEvent">
    <Exec Command="echo Hello" />
  </Target>

  <Target Name="PreBuild2" BeforeTargets="PreBuildEvent">
    <Exec Command="echo World" />
  </Target>
```

We will only display the first:
image

Modifying it (and saving) updates the first target:
```xml
  <Target Name="PreBuild1" BeforeTargets="PreBuildEvent">
    <Exec Command="echo Hello, " />
  </Target>

  <Target Name="PreBuild2" BeforeTargets="PreBuildEvent">
    <Exec Command="echo World" />
  </Target>
```

Fully deleting the text (and saving) causes the first target (PreBuild1) to be deleted from the csproj file.

```xml
  <Target Name="PreBuild2" BeforeTargets="PreBuildEvent">
    <Exec Command="echo World" />
  </Target>
```

Then, if you modify the text box again (and save), we now end up modifying the second target:

  <Target Name="PreBuild2" BeforeTargets="PreBuildEvent">
    <Exec Command="echo a" />
  </Target>

This results in us overwriting user targets (without them being aware that it switched the target is was matched against).

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+                    }
+                }
+
+                SetParameter(projectXml, unevaluatedPropertyValue);
+            }
+
+            private (bool success, string property) TryGetFromProperty(ProjectRootElement projectXml)
+            {
+                // Choose last or default as that matches the evaluation order for MSBuild
+                var property = projectXml.Properties
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs]: Don't use the construction model to get properties - if the property was inside a condition or if it's for a certain configuration then this will be wrong. Instead, import ProjectPropertiex and get the PreBuildEvent property from GeneralBrowseObject - however you should set it's SourceOfDefaultValue=BeforeContext [here](https://github.com/dotnet/project-system/blob/6edd3dae1a9d8435d28d22971a73c7a5000c36b7/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/GeneralBrowseObject.xaml#L72) so that it doesn't give you the final value.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+                    return commandText;
+                }
+
+                return null; // exec task as written in the project file is invalid, we should be resilient to this case.
+            }
+
+            private bool TrySetProperty(string unevaluatedPropertyValue, ProjectRootElement projectXml)
+            {
+                // Choose last or default as that matches the evaluation order for MSBuild
+                var property = projectXml.Properties
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs]: Same as above - go through ProjectProperties.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+                if (targetNames.Contains(targetName))
+                {
+                    targetName = FindNonCollidingName(targetName, targetNames);
+                }
+
+                return targetName;
+
+            }
+
+            private string FindNonCollidingName(string buildEvent, HashSet<string> targetNames)
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs]: Let`s not do this - if the exact name doesn't match then that should be an error - this just complicates things without much value.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs (hunk 19 above)
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs]: Ah actually just noticed that you are implementing an intercepter for this property anyway. So just check if evaluatedProeprtyValue [here](https://github.com/dotnet/project-system/pull/2367/files#diff-df8f51b25a07aeabf5c8cb6e9b93f764R26) is non-null. You can avoid this whole TryGet\SetFromProperty thing. You'd still need to change SourceOfDefaultValue.

**Comment by srivatsn** [comment]
Tagging @MattGertz for Preview4

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Utilities/StringExtensions.cs [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            var stringReader = new System.IO.StringReader(@string);
+            var xmlReader = new XmlTextReader(stringReader);
+            var root = ProjectRootElement.Create(xmlReader);
+            return root;
+        }
+
+        public static string SaveAndGetChanges(this ProjectRootElement root)
+        {
+            var tempFile = Path.GetTempFileName();
```
Comment (vshapenko) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Utilities/StringExtensions.cs]: Why cant we simply use root.Save(TextWriter) here?

**Comment by srivatsn** [comment]
@MattGertz tagging again in case you missed this - this is one of the top vsfeedback issues we've had.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Utilities/StringExtensions.cs (hunk 22 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Utilities/StringExtensions.cs]: @vshapenko saving to a string will cause a different code path to be used that writes out the file differently. I am writing to disk here because I want the same behavior in the unit tests and what MSBuild actually does. I can look for other ways around this and revisit this is a subsequent PR
